### PR TITLE
Enhancements for VSCode Grammar

### PR DESCRIPTION
**Pull request by dusty-phillips**
Some of these changes are debatable,
so I just wanted to put this out there as
a suggestion:

 *  ~~For the `<action> line` commands, I made the word `line`
 no longer optional.
**I did this because dragon was often interpreting
 `deli two` as `delete two`
 and unexpectedly deleted the second line
 instead of two characters.**~~
 *  move "show terminal" from
 non-ccr to ccr.
** I did this because I very frequently want to say things like
 `terminal cargo build`**
 to run a command without waiting
 for the terminal to be expected.
 * ~~Add alternative `[manpower]` to the `command palette` command.
**On my bad microphone, `command palette` was interpreted
 as `manpower`**
 too often.~~
 *  Add `[<text>]` to the `command palette`
 command
**so you can say the command
 you want in a single statement**
 *  Add "Open project" command to open a recent folder.
 **Note: The `[<text>]` in `Open project` doesn't always get picked up
 and gets inserted into the current file
 instead. I don't know how to fix this.**

## How Has This Been Tested

I've been using these commands for quite a while, but my keybindings for the open file/recent/commandpallette are different from the default, so it might be worth double checking them.

## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue or bug)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Renamed existing command phrases (we discourage this without a strong rationale).

## Checklist

- [x] I have read the CONTRIBUTING document.
- [x] My code follows the code style of this project.
- [x] I have checked that my code does not duplicate functionality elsewhere in Caster.
- [x] I have checked for and utilized existing command phrases from within Caster (delete if not applicable). 
- [x] My code implements all the features I wish to merge in this pull request.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests pass.

## Maintainer/Reviewer Checklist

- [ ] Basic functionality has been tested and works as claimed.
- [ ] New documentation is clear and complete.
- [ ] Code is clear and readable.

### CODE REVIEW

**Comment by LexiconCode** [comment]
For a change like this I would recommend using simplified transformers and adding to word.txt. It's careful balance between our personal setups and what everyone else uses. Otherwise people have their grammars change when they update and get unexpected behavior.

>     * Add alternative `[manpower]` to the `command palette` command.
>       **On my bad microphone, `command palette` was interpreted
>       as `manpower`**


>     * For the `<action> line` commands, I made the word `line`
>       no longer optional.
>       **I did this because dragon was often interpreting
>       `deli two` as `delete two`
>       and unexpectedly deleted the second line
>       instead of two characters.**

Snagging this from the documentation that I'm updating.
​	What Transformers do is expose pairs of rules to the user, allowing the user to change the rules' mapping as they see fit. This is recommended due to caster grammars updating where user directory grammars. Great for small changes across many grammars without editing grammar source code or copying premade rules.  Transformers are applied during boot time. Details are in the CCR section of the documentation.

​	Though rule transformersare very powerful, setting one up for the first time is not trivial. Therefore, a simplified method for common use cases has been created. Create the file `Local\caster\transforms\words.txt` and it will be read at boot time, and a rule filter created from it and added to Caster's list of rule filters. The following is an example `words.txt`:

```
<<<SPEC>>>
shock -> earthquake
<<<NOT_SPECS>>>
sauce -> up
dunce -> down
lease -> left
```

​	This `words.txt` will create a rule filter which goes through **all** Caster rules at boot time and replaces the word "shock" in any spec with the word "earthquake". It will also replace "sauce", "dunce" and "lease" with "up", "down", and "left" in extras and defaults in any rule. The triple-angle-brackets indicate mode changes. Valid modes are `SPEC` (for specs only), `EXTRA`, `DEFAULT`, `NOT_SPECS` (for extras and defaults, but not specs), and `ANY` (for specs, extras, and defaults). The default mode is `ANY`.

### 

As for the rest of the suggestions you recommended I think could be added without trouble and increase utility of the VS code grammar!

**Comment by dusty-phillips** [comment]
What would the transformer look like for those two commands? I only knew about using them for mapping individual words.

I'll remove those two from my pull request, so you can incorporate the rest. Do you have any ideas for fixing the race condition I mentioned in the `open project` command?

**Comment by LexiconCode** [comment]
@dusty-phillips 
Come to think of it that's not in the documentation which should be added that words a replaceable but entire commands as well.
```
<<<SPEC>>>
[open] command palette -> [open] (command palette | manpower)
<action> [line] <ln1> [by <ln2>] -> <action> line <ln1> [by <ln2>]
```
Let me know if you have issues with the second one not working. I suspect it might be related to having `<>` characters.

**Comment by dusty-phillips** [comment]
The pause trick worked, but neither of the transformer commands did. Since I was working on my branch, i ended up having angle brackets in both commands:

```
<<<SPEC>>>
[open] command palette [<text>] -> [open] (command palette | manpower) [<text>]
<action> [line] <ln1> [by <ln2>] -> <action> line <ln1> [by <ln2>]
```

However there were no error messages; it just didn't work. It's not a major issue for me, since I'm managing some customizations on top of the master branch, to support some of my custom key bindings anyway. But I thought I'd report the issue either way.

**Comment by LexiconCode** [comment]
> 
> 
> The pause trick worked, but neither of the transformer commands did. Since I was working on my branch, i ended up having angle brackets in both commands:
> 
> ```
> <<<SPEC>>>
> [open] command palette [<text>] -> [open] (command palette | manpower) [<text>]
> <action> [line] <ln1> [by <ln2>] -> <action> line <ln1> [by <ln2>]
> ```
> 
> However there were no error messages; it just didn't work. It's not a major issue for me, since I'm managing some customizations on top of the master branch, to support some of my custom key bindings anyway. But I thought I'd report the issue either way.

Could you try testing the process out on a complete command with out brackets?

**Comment by dusty-phillips** [comment]
Added `Pause` to the pull request.

I confirmed the mapping commands without angle brackets works:

```
<<<SPEC>>>
[toggle] full screen -> full steam ahead
```

File: castervoice/rules/apps/editor/vscode_rules/vscode2.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ class VSCodeNonCcrRule(MappingRule):
         # Moving around a file
         "[(go to | jump | jump to)] line <n>":
             R(Key("c-g") + Text("%(n)d") + Key("enter")),
-        "<action> [line] <ln1> [by <ln2>]":
+        "<action> line <ln1> [by <ln2>]":
```
Comment (LexiconCode) [inline castervoice/rules/apps/editor/vscode_rules/vscode2.py]: line is not optional -> [line]

File: castervoice/rules/apps/editor/vscode_rules/vscode2.py [hunk 2]
```diff
@@ -75,10 +75,12 @@ class VSCodeNonCcrRule(MappingRule):
         # File Management
         "copy path":
             R(Key("c-k, p")),
-        "[open] command palette":
-            R(Key("cs-p"), rdescript="VS Code: Command Palette"),
+        "[open] command palette [<text>]":
```
Comment (LexiconCode) [inline castervoice/rules/apps/editor/vscode_rules/vscode2.py]: We might as well add `Pause("30")` can at the same raising condition as open project.

File: castervoice/rules/apps/editor/vscode_rules/vscode2.py (hunk 1 above)
Comment (dusty-phillips) [reply castervoice/rules/apps/editor/vscode_rules/vscode2.py]: Sorry about that. I thought I'd already removed it.

**Pull request merged**